Separar download sat

ppaez opened the pull request:
Esta rama derivada de develop es para realizar el issue #25.  Empieza con agregados a la documentación que son necesarios, se agregarán más cambios en código esta tarde al final de la sesión de TDD en LinuxCabal.

Una vez que esté completa, esta rama podría mezclarse en un release-0.2.7 o subsequente.

Commit:
Merge branch 'hotfix-0.2.5' into develop

Se agrega soporte para obtener el detalle de conceptos en los reportes, por ejemplo:

{emisor_rfc}|{serie}{folio}|{noIdentificacion}|{descripcion}|{unidad}|{cantidad}|{valorUnitario}|{importe}

Si agrega cualquier campo del detalle, se omite la validación de los sellos y del SAT

mimex committed:
Merge branch 'hotfix-0.2.6' into develop

ppaez committed:
Empezar la referencia de los módulos

- Habilitar la extensión autodoc
- Crear la sección Referencia

ppaez committed:
Explicar la descarga de facturas del SAT

- Qué código está involucrado
- Los pasos generales

ppaez committed:
Explicar los principales módulos de admin-cfdi

- Por ahora solo los que tienen que ver con
  la descarga de CFDIs

ppaez committed:
Crear DescargaSAT

- Es copia sin refactorizar de _download_sat()
- Queda en pyutil.py como las demás herramientas

ppaez committed:
Usar DescargaSAT en script

- Crear descarga.py como script
  para ejecutar desde consola
- Crear pwd.sample como plantilla
  del archivo de credenciales

ppaez commented:
Ya se agregaron cambios en esta rama, que crean DescargaSAT y descarga.py.  Si quieren probar descarga.py, es un script de consola que usa DescargaSAT.

ppaez committed:
Crear método _download_sat()

- Alineando con la definición propuesta en
  el issue #25

ppaez committed:
Agregar parámetros opcionales para la línea de comando

- Ayuda con -h
- Se ofrecen valores predeterminados

ppaez committed:
Usar ambos parámetros de línea de comando

- La carpeta destino tiene que ser una ruta
  absoluta.  Si es relativa, el navegador
  la ignora y utiliza su valor predeterminado
  de descargas :(

ppaez committed:
Explicar validate_download_sat()

- Las validaciones que hace
- El diccionario que construye
- La tupla que regresa

ppaez committed:
Explicar los valores de data

ppaez committed:
Correcciones

ppaez committed:
Agregar parámetros opcionales año, mes y día

- Los valores predeterminados igual que
  en la aplicación admincfdi

ppaez committed:
Corrección

ppaez committed:
Agregar parámetros opcionales facturas-emitidas y uuid

- El uso de uuid cambia type_search a 1

ppaez committed:
Completar el mock de la aplicación

ppaez committed:
Agregar parámetro opcional rfc-emisor

ppaez committed:
Renombrar self a app

- Se preserva self para referir a
  _download_sat_month() que se va
  a agregar ahora
- Ya podemos refactorizar _download_sat()
  porque podemos probar funcionalmente
  todas las opciones de búsqueda

ppaez committed:
Crear _download_sat_month()

- Es copia sin refactorizar de _download_sat_month()
- Como método de DescargaSAT

ppaez committed:
Agregar parámetro opcional mes-completo

ppaez committed:
Completar el mock de la aplicación

ppaez commented:
Se agregaron más cambios: descarga.py ya tiene opciones de línea de comando para todos los tipos de búsquedas, pueden listar las opciones así:

```
(env34) pp@g6:~/admin-cfdi$ py descarga.py -h
usage: descarga.py [-h] [--archivo-de-credenciales ARCHIVO_DE_CREDENCIALES]
                   [--carpeta-destino CARPETA_DESTINO] [--facturas-emitidas]
                   [--uuid UUID] [--rfc-emisor RFC_EMISOR] [--año AÑO]
                   [--mes MES] [--día DÍA] [--mes-completo]

Descarga CFDIs del SAT a una carpeta local

optional arguments:
  -h, --help            show this help message and exit
  --archivo-de-credenciales ARCHIVO_DE_CREDENCIALES
                        Archivo con credenciales para el SAT. RFC y CIEC en el
                        primer renglón y separadas por un espacio. El
                        predeterminado es pwd
  --carpeta-destino CARPETA_DESTINO
                        Carpeta local para guardar los CFDIs descargados El
                        predeterminado es cfdi
  --facturas-emitidas   Descargar facturas emitidas. Por omisión se descargan
                        facturas recibidas
  --uuid UUID           UUID. Por omisión no se usa en la búsqueda. Esta
                        opción tiene precedencia sobre las demás opciones de
                        búsqueda.
  --rfc-emisor RFC_EMISOR
                        RFC del emisor. Por omisión no se usa en la búsqueda.
  --año AÑO             Año. El valor por omisión es el año en curso
  --mes MES             Mes. El valor por omisión es el mes en curso
  --día DÍA             Día. El valor por omisión es '00', significa no usar
                        el día en la búsqueda
  --mes-completo        Mes completo por día. Por omisión no se usa en la
                        búsqueda.
```

Si quieren probar descarga.py, hagan pull de esta rama en mi fork.  Lo he estado usando para probar funcionalmente los cambios y está muy padre hacer las descagas desde bash ;)

Espero que ya pronto lleguemos al cambio para que admincfdi.py use directamente DescargaSAT

ppaez committed:
Agrupar el bloque main

ppaez committed:
Mover app afuera de main

- Se hará innecesaria mediante refactorización
  en DescargaSAT

mr-E commented:
Buen día

``` quote
...probar descarga.py, hagan pull de esta rama en mi fork...
```

Quiero probar la descarga.py... ¿como hacer el pull de la rama?
Todavia no termino de auto-capacitarme en GitHub. :$

ppaez commented:
@mr-E: Debería decir hacer _checkout_, en lugar pull.  Si estás dentro del repositorio admin-cfdi en tu PC, agrega una referencia remota a mi fork del proyecto así:

```
  git remote add ppaez https://github.com/ppaez/admin-cfdi.git
```

Actualiza las referencias en tu repositorio:

```
  git fetch ppaez
```

Ahora sí, haz checkout de la rama separar-dowload-sat:

```
git checkout separar-download-sat
Branch separar-download-sat set up to track remote branch separar-download-sat from ppaez.
Switched to a new branch 'separar-download-sat'
```

Estando en esta rama, deberás ver descarga.py en tu repositorio.

Algo que no he resuelto aún, es porqué FireFox ignora el valor de **--carpeta-destino** si es una ruta relativa... usa una ruta absoluta como ~/cfdi, y crea esa carpeta antes de correr descarga.py.

ghost commented:
Cuando hago
`git checkout separar-download-sat`
Obtengo

```
error: Your local changes to the following files would be overwritten by checkout:
    admincfdi.py
Please, commit your changes or stash them before you can switch branches.
Aborting
```

Creo recordar que hay que pasarle rutas abosultas a Firefox...

ghost commented:
Eran unos cambios que había hecho...

probando descarga.py

ghost commented:
Prueba exitosa, quedo muy padre...

ppaez committed:
Renombrar self a app

ppaez committed:
Utilizar values.Global

- Se usan el atributo SAT, page_init y page_cfdi
- Remover del mock de aplication

ppaez committed:
Utilizar pyutil.Util

- Usar get_dates, get_days, sleep
- Remover del mock de aplication

ppaez committed:
Agregar parámetro opcional download_callback

- El atributo progress es llamado para
  proporcionar valor y máximo a un control
  gráfico de tipo barra de progreso
- Por omisión se envían valor y máximo
  a stdout
- Remover pb del mock de aplication

ppaez committed:
Agregar parámetro opcional status_callback

- El atributo status es llamado para
  enviar mensajes
- Por omisión los mensajes se envían
  a stdout
- Remover del mock de aplication

ppaez committed:
Remover parámetro app

- Remover el mock

mr-E commented:
Intente probar "descarga.py" pero me falta:

``` code
from unittest.mock import Mock
from unittest.mock import MagicMock
```

Usando pip intente instalar "unittest", "unittest2" y "mock", pero no funciona.
Por ejemplo

``` code
(virt)...$ pip install unittest
```

No lo encuentra y menciona en el log que va a ignorar unittest2 :

``` code
(cfdadmin)mre@ubuntu:~/virt/cfdadmin/admin-cfdi-separar-download-sat$ pip install unittest
Downloading/unpacking unittest
  Could not find any downloads that satisfy the requirement unittest
Cleaning up...
No distributions at all found for unittest
Storing complete log in /tmp/tmpd6bskp:
....
../virt/cfdadmin/bin/pip run on Tue Mar  3 08:44:05 2015
Downloading/unpacking unittest

  Getting page https://pypi.python.org/simple/unittest/
  URLs to search for versions for unittest:
  * https://pypi.python.org/simple/unittest/
  Analyzing links from page https://pypi.python.org/simple/unittest/
    Skipping link https://pypi.python.org/packages/source/u/unittest/unittest2-0.0.0.tar.gz#md5=2a416384f814604ed854d40f352f1191 (from https://pypi.python.org/simple/unittest/); wrong project name (not unittest)
  Could not find any downloads that satisfy the requirement unittest

Cleaning up...
```

ppaez commented:
unittest es parte de la librería estandar, pero mock fue introducido en Python 3.3, eso es lo que posiblemente te está faltando.  Si tienes Python 3.2 por ejemplo, podrías intentar

```
 pip install mock
```

y modifca esos dos import para que digan:

```
 from mock import Mock
 from mock import MagicMock
```

Pero hay una opción mejor: esta mañana subí cambios que removieron el uso de unittest.mock en descarga.py :)  Simplemente actualiza tu repositorio así:

```
 git pull ppaez separar-download-sat
```

mr-E commented:
+1 Gracias!

ppaez committed:
admincfdi usa DescargaSAT

- Agregar los métodos msg_user() y progress()
  como wrapper a los widgets
- Usar estos métodos en los parámetros opcionales
  status_callback y download_callback

ppaez committed:
Remover descarga de admincfdi

- _download_sat() y _download_sat_month()
  ya están incluidas en pyutil.DescargaSAT

ppaez commented:
La mezcla a master se deshizo, no era lo que se deseaba y no nos dimos cuenta, y se rehizo la mezcla a la rama develop.
